packages/plugin-essentials/sources/commands/plugin/import/sources.ts

```diff
@@ -0,0 +1,99 @@
+import {BaseCommand}                                                    from '@yarnpkg/cli';
+import {structUtils}                                                    from '@yarnpkg/core';
+import {Configuration, MessageName, Project, ReportError, StreamReport} from '@yarnpkg/core';
+import {PortablePath, npath, ppath, xfs, Filename}                      from '@yarnpkg/fslib';
+import {Command, Usage}                                                 from 'clipanion';
+import {tmpdir}                                                         from 'os';
+
+import {prepareRepo, runWorkflow}                                       from '../../set/version/sources';
+import {savePlugin}                                                     from '../import';
+
+import {getAvailablePlugins}                                            from '../list';
+
+const BUILD_WORKFLOW = ({pluginName, noMinify}: {noMinify: boolean, pluginName: string}, target: PortablePath) => [
+  [`yarn`, `build:${pluginName}`, ...noMinify ? [`--no-minify`] : [], `|`],
+];
+
+// eslint-disable-next-line arca/no-default-export
+export default class PluginDlSourcesCommand extends BaseCommand {
+  @Command.String()
+  name!: string;
+
+  @Command.String(`--path`)
+  installPath?: string;
+
+  @Command.String(`--repository`)
+  repository: string = `https://github.com/yarnpkg/berry.git`;
+
+  @Command.String(`--branch`)
+  branch: string = `master`;
+
+  @Command.Boolean(`--no-minify`)
+  noMinify: boolean = false;
+
+  @Command.Boolean(`-f,--force`)
+  force: boolean = false;
+
+  static usage: Usage = Command.Usage({
+    category: `Plugin-related commands`,
+    description: `build a plugin from sources`,
+    details: `
+      This command clones the Yarn repository into a temporary folder, builds the specified contrib plugin and updates the configuration to reference it in further CLI invocations.
+
+      The plugin can be referenced by name.
+    `,
+    examples: [[
+      `Build and activate the "@yarnpkg/plugin-exec" plugin`,
+      `$0 plugin import from sources @yarnpkg/plugin-exec`,
+    ], [
+      `Build and activate the "@yarnpkg/plugin-exec" plugin (shorthand)`,
+      `$0 plugin import from sources exec`,
+    ]],
+  });
+
+  @Command.Path(`plugin`, `import`, `from`, `sources`)
+  async execute() {
+    const configuration = await Configuration.find(this.context.cwd, this.context.plugins);
+
+    const target = typeof this.installPath !== `undefined`
+      ? ppath.resolve(this.context.cwd, npath.toPortablePath(this.installPath))
+      : ppath.resolve(npath.toPortablePath(tmpdir()), `yarnpkg-sources` as Filename);
```

Comment:
That makes me thing that perhaps we should add a subfolder component based on the hash of the repository? Otherwise if you download a plugin from a different repo it'll forget to refetch it.